Doc/library/importlib.rst

```diff
@@ -795,6 +797,28 @@ ABC hierarchy::
         itself does not end in ``__init__``.
 
 
+.. class:: Traversable
+
+    An object with a subset of pathlib.Path methods suitable for
+    traversing directories and opening files.
+
+    .. versionadded:: 3.9
+
+
+.. class:: TraversableReader
+
+    An abstract base class for resource readers capable of serving
+    the ``files`` interface. Subclasses ResourceReader and provides
+    concrete implementations of the ResourceReader's abstract
+    methods. Therefore, any loader supplying TraversableReader
+    also supplies ResourceReader.
+
+    Loaders that wish to support resource reading are expected to
```

Comment:
@brettcannon , @warsaw  In earlier versions, a loader was expected to supply a `get_resource_reader()` which return a `ResourceReader`. In my design, I avoid that level of indirection and ask the `Loader` to subclass `TraversableResources` and implement `files()` directly. The implications of this are:

- There's no intermediate 'reader' object. The `loader` is the `reader`.
- By deriving from `TraversableResources`, the `loader` is also the `ResourceReader`.

Was there a compelling reason why loaders would return ResourceReaders instead of simply being ResourceReaders themselves? Is there any case where a `Loader` would not know the context of the package for which it loads?

Reply:
Because we didn't want to edit zipimport to subclass something because neither of us had the stomach to write the C code. 😁  Plus we had no issue with delegation to a different object.

Reply:
Okay, that helps. Thanks for the clarification. I'll similarly want to write as thin shim on the loaders for `files()` support.